File: crates/pg_lens_web/frontend/src/blocks-lens.ts
// Blocks & Locks lens (v0.17): wait-for dependency tree and active locks table.
//
// Pure TypeScript rendering over `DbSnapshot.blocking_tree` and
// `DbSnapshot.active_locks`, mirroring the TUI's BlocksLens layout.

import type { ActiveLockRow, BlockTreeNode } from "./types";
import { humanDuration } from "./format.ts";

export interface BlocksLens {
  update(tree: BlockTreeNode[] | null, locks: ActiveLockRow[] | null): void;
}

export function lockStatusBadge(granted: boolean): { text: string; className: string } {
  return granted
    ? { text: "GRANT", className: "badge badge-ok" }
    : { text: "WAIT", className: "badge badge-bad" };
}

export function blockNodeBadge(node: BlockTreeNode): { text: string; className: string } | null {
  if (node.is_deadlock) {
    return { text: "DEADLOCK", className: "badge badge-bad" };
  }
  if (node.is_root) {
    const s = node.num_descendants === 1 ? "" : "s";
    return {
      text: `ROOT BLOCKER (${node.num_descendants} waiter${s})`,
      className: "badge badge-bad",
    };
  }
  return null;
}

export function countTreeNodes(nodes: BlockTreeNode[]): number {
  let count = 0;
  for (const n of nodes) {
    count += 1 + countTreeNodes(n.children || []);
  }
  return count;
}

export function initBlocksLens(panel: HTMLElement): BlocksLens {
  const treeContainer = panel.querySelector<HTMLElement>("#blocks-tree-container");
  const locksTbody = panel.querySelector<HTMLTableSectionElement>("#blocks-locks tbody");

  return {
    update(tree: BlockTreeNode[] | null, locks: ActiveLockRow[] | null): void {
      if (treeContainer) {
        renderTree(treeContainer, tree);
      }
      if (locksTbody) {
        renderLocks(locksTbody, locks);
      }
    },
  };
}

export function renderTree(container: HTMLElement, tree: BlockTreeNode[] | null): void {
  container.replaceChildren();

  if (!tree || tree.length === 0) {
    const empty = document.createElement("p");
    empty.className = "placeholder";
    empty.textContent = "No blocked sessions detected — all queries running freely.";
    container.appendChild(empty);
    return;
  }

  const list = document.createElement("div");
  list.className = "blocks-tree-list";

  function renderNodes(nodes: BlockTreeNode[], depth: number): void {
    for (const node of nodes) {
      const row = document.createElement("div");
      row.className = "blocks-tree-row";
      row.style.setProperty("--depth", String(depth));

      const header = document.createElement("div");
      header.className = "blocks-tree-header";

      const pidSpan = document.createElement("span");
      pidSpan.className = "blocks-tree-pid";
      pidSpan.textContent = `PID ${node.pid}`;
      header.appendChild(pidSpan);

      const badgeInfo = blockNodeBadge(node);
      if (badgeInfo) {
        const badge = document.createElement("span");
        badge.className = badgeInfo.className;
        badge.textContent = badgeInfo.text;
        header.appendChild(badge);
      }

      const userApp = document.createElement("span");
      userApp.className = "blocks-tree-user";
      userApp.textContent = `${node.usename}@${node.application_name}`;
      header.appendChild(userApp);

      const state = document.createElement("span");
      state.className = `blocks-tree-state ${node.state === "idle in transaction" ? "warn" : ""}`;
      state.textContent = node.state;
      header.appendChild(state);

      if (node.mode) {
        const lockInfo = document.createElement("span");
        lockInfo.className = "blocks-tree-lock";
        lockInfo.textContent = node.relation ? `${node.mode} on ${node.relation}` : node.mode;
        header.appendChild(lockInfo);
      }

      const dur = document.createElement("span");
      dur.className = "blocks-tree-dur";
      dur.textContent = humanDuration(node.duration_secs);
      header.appendChild(dur);

      row.appendChild(header);

      if (node.query) {
        const q = document.createElement("div");
        q.className = "blocks-tree-query sql-text";
        q.textContent = node.query;
        row.appendChild(q);
      }

      list.appendChild(row);

      if (node.children && node.children.length > 0) {
        renderNodes(node.children, depth + 1);
      }
    }
  }

  renderNodes(tree, 0);
  container.appendChild(list);
}

export function renderLocks(tbody: HTMLTableSectionElement, locks: ActiveLockRow[] | null): void {
  tbody.replaceChildren();

  if (!locks || locks.length === 0) {
    const tr = document.createElement("tr");
    const td = document.createElement("td");
    td.colSpan = 8;
    td.className = "placeholder";
    td.textContent = "No active relation or transaction locks in database.";
    tr.appendChild(td);
    tbody.appendChild(tr);
    return;
  }

  for (const lock of locks) {
    const tr = document.createElement("tr");
    if (!lock.granted) {
      tr.className = "lock-waiting-row";
    }

    const tdPid = document.createElement("td");
    tdPid.textContent = String(lock.pid);
    tdPid.className = "num";

    const tdType = document.createElement("td");
    tdType.textContent = lock.locktype;

    const tdRel = document.createElement("td");
    tdRel.textContent = lock.relation
      ? (lock.schema && lock.schema !== "public" ? `${lock.schema}.${lock.relation}` : lock.relation)
      : "—";

    const tdMode = document.createElement("td");
    tdMode.textContent = lock.mode;

    const tdStatus = document.createElement("td");
    const badgeInfo = lockStatusBadge(lock.granted);
    const statusBadge = document.createElement("span");
    statusBadge.className = badgeInfo.className;
    statusBadge.textContent = badgeInfo.text;
    tdStatus.appendChild(statusBadge);

    const tdAge = document.createElement("td");
    tdAge.textContent = humanDuration(lock.duration_secs);
    tdAge.className = "num";

    const tdUser = document.createElement("td");
    tdUser.textContent = lock.usename || "—";

    const tdQuery = document.createElement("td");
    tdQuery.className = "query-cell sql-text";
    tdQuery.textContent = lock.query.replace(/\n/g, " ");

    tr.append(tdPid, tdType, tdRel, tdMode, tdStatus, tdAge, tdUser, tdQuery);
    tbody.appendChild(tr);
  }
}

File: crates/pg_lens_tui/src/ui/blocks_lens.rs
//! Blocks & Locks Lens (v0.17): hierarchical wait-for tree (who blocks whom)
//! and active database lock inventory (`pg_locks`).
//!
//! Split-pane layout:
//! - Top: Hierarchical Blocking Tree with Root Blockers, deadlock detection,
//!   and dependency depth.
//! - Bottom: Active Locks Table showing granted vs waiting lock modes, relations,
//!   and query snippets.

use pg_lens_core::BlockTreeNode;
use ratatui::{
    Frame,
    layout::{Constraint, Layout, Rect},
    style::{Color, Modifier, Style, Stylize},
    text::{Line, Span},
    widgets::{Block, Borders, Clear, Paragraph, Row, Table, Wrap},
};

use crate::app::{App, BlocksPane};
use crate::ui::{format, sql};

pub struct FlattenedNode<'a> {
    pub depth: usize,
    pub node: &'a BlockTreeNode,
    pub is_last_child: bool,
}

pub fn flatten_tree<'a>(nodes: &'a [BlockTreeNode]) -> Vec<FlattenedNode<'a>> {
    let mut out = Vec::new();
    fn walk<'a>(nodes: &'a [BlockTreeNode], depth: usize, out: &mut Vec<FlattenedNode<'a>>) {
        let count = nodes.len();
        for (i, node) in nodes.iter().enumerate() {
            out.push(FlattenedNode {
                depth,
                node,
                is_last_child: i + 1 == count,
            });
            walk(&node.children, depth + 1, out);
        }
    }
    walk(nodes, 0, &mut out);
    out
}

pub fn draw(app: &mut App, frame: &mut Frame, area: Rect) {
    let [tree_area, locks_area] = Layout::vertical([
        Constraint::Percentage(55),
        Constraint::Percentage(45),
    ])
    .areas(area);

    draw_tree(app, frame, tree_area);
    draw_locks(app, frame, locks_area);

    if app.detail_open {
        draw_detail(app, frame, area);
    }
}

fn draw_tree(app: &mut App, frame: &mut Frame, area: Rect) {
    let is_active_pane = app.blocks_active_pane == BlocksPane::Tree;
    let border_style = if is_active_pane {
        Style::new().fg(Color::Yellow)
    } else {
        Style::new().dim()
    };

    let tree_data = app.snapshot.blocking_tree.as_deref().unwrap_or(&[]);
    let flattened = flatten_tree(tree_data);

    let title = format!(
        " Blocking Tree ({} roots, {} blocked) ",
        tree_data.len(),
        flattened.iter().filter(|f| !f.node.is_root).count()
    );

    let block = Block::default()
        .borders(Borders::ALL)
        .border_style(border_style)
        .title(Span::styled(title, Style::new().bold()));

    if flattened.is_empty() {
        let empty = Paragraph::new(Line::from("  No blocked sessions detected \u{2014} all queries running freely.").dim())
            .block(block);
        frame.render_widget(empty, area);
        return;
    }

    let rows: Vec<Row> = flattened
        .iter()
        .map(|item| {
            let n = item.node;
            let mut prefix = String::new();
            if item.depth > 0 {
                for _ in 0..(item.depth - 1) {
                    prefix.push_str("   ");
                }
                if item.is_last_child {
                    prefix.push_str(" └─ ");
                } else {
                    prefix.push_str(" ├─ ");
                }
            }

            let pid_span = if n.is_deadlock {
                Span::styled(format!("{prefix}[DEADLOCK] PID {}", n.pid), Style::new().fg(Color::Red).bold())
            } else if n.is_root {
                Span::styled(
                    format!("{prefix}[ROOT BLOCKER] PID {} ({} waiters)", n.pid, n.num_descendants),
                    Style::new().fg(Color::Red).bold(),
                )
            } else {
                Span::styled(format!("{prefix}PID {}", n.pid), Style::new().fg(Color::Yellow))
            };

            let user_app = format!("{}@{}", n.usename, n.application_name);
            let state_style = if n.state == "idle in transaction" {
                Style::new().fg(Color::Yellow).bold()
            } else if n.state == "active" {
                Style::new().fg(Color::Green)
            } else {
                Style::new().dim()
            };

            let lock_info = if let (Some(m), Some(r)) = (&n.mode, &n.relation) {
                if !r.is_empty() {
                    format!("{m} on {r}")
                } else {
                    m.clone()
                }
            } else if let Some(m) = &n.mode {
                m.clone()
            } else {
                "-".to_string()
            };

            let dur = format::human_duration(n.duration_secs);

            Row::new(vec![
                Line::from(pid_span),
                Line::from(Span::styled(user_app, Style::new().dim())),
                Line::from(Span::styled(n.state.clone(), state_style)),
                Line::from(Span::styled(lock_info, Style::new().cyan())),
                Line::from(Span::styled(dur, Style::new().dim())),
                Line::from(Span::raw(n.query.replace('\n', " "))),
            ])
        })
        .collect();

    let widths = [
        Constraint::Length(34),
        Constraint::Length(18),
        Constraint::Length(20),
        Constraint::Length(24),
        Constraint::Length(8),
        Constraint::Min(20),
    ];

    let highlight_style = if is_active_pane {
        Style::new().add_modifier(Modifier::REVERSED).bold()
    } else {
        Style::new().bg(Color::DarkGray)
    };

    let table = Table::new(rows, widths)
        .block(block)
        .header(
            Row::new(vec![
                "Session / Hierarchy",
                "User@App",
                "State",
                "Awaiting Lock",
                "Age",
                "Query",
            ])
            .style(Style::new().bold().dim()),
        )
        .row_highlight_style(highlight_style);

    frame.render_stateful_widget(table, area, &mut app.blocks_tree_state);
}

fn draw_locks(app: &mut App, frame: &mut Frame, area: Rect) {
    let is_active_pane = app.blocks_active_pane == BlocksPane::Locks;
    let border_style = if is_active_pane {
        Style::new().fg(Color::Yellow)
    } else {
        Style::new().dim()
    };

    let locks = app.snapshot.active_locks.as_deref().unwrap_or(&[]);

    let title = format!(" Active Locks ({} entries) ", locks.len());

    let block = Block::default()
        .borders(Borders::ALL)
        .border_style(border_style)
        .title(Span::styled(title, Style::new().bold()));

    if locks.is_empty() {
        let empty = Paragraph::new(Line::from("  No active relation or transaction locks in database.").dim())
            .block(block);
        frame.render_widget(empty, area);
        return;
    }

    let rows: Vec<Row> = locks
        .iter()
        .map(|l| {
            let pid_span = Span::styled(l.pid.to_string(), Style::new().bold());

            let (status_span, row_style) = if !l.granted {
                (
                    Span::styled("WAIT", Style::new().fg(Color::Red).bold()),
                    Style::new().fg(Color::LightRed),
                )
            } else {
                (
                    Span::styled("GRANT", Style::new().fg(Color::Green)),
                    Style::new(),
                )
            };

            let rel = if !l.relation.is_empty() {
                if !l.schema.is_empty() && l.schema != "public" {
                    format!("{}.{}", l.schema, l.relation)
                } else {
                    l.relation.clone()
                }
            } else {
                "-".to_string()
            };

            let dur = format::human_duration(l.duration_secs);

            Row::new(vec![
                Line::from(pid_span),
                Line::from(Span::raw(l.locktype.clone())),
                Line::from(Span::styled(rel, Style::new().cyan())),
                Line::from(Span::raw(l.mode.clone())),
                Line::from(status_span),
                Line::from(Span::styled(dur, Style::new().dim())),
                Line::from(Span::styled(l.usename.clone(), Style::new().dim())),
                Line::from(Span::raw(l.query.replace('\n', " "))),
            ])
            .style(row_style)
        })
        .collect();

    let widths = [
        Constraint::Length(8),
        Constraint::Length(14),
        Constraint::Length(22),
        Constraint::Length(24),
        Constraint::Length(8),
        Constraint::Length(8),
        Constraint::Length(12),
        Constraint::Min(20),
    ];

    let highlight_style = if is_active_pane {
        Style::new().add_modifier(Modifier::REVERSED).bold()
    } else {
        Style::new().bg(Color::DarkGray)
    };

    let table = Table::new(rows, widths)
        .block(block)
        .header(
            Row::new(vec![
                "PID",
                "Type",
                "Relation",
                "Mode",
                "Status",
                "Age",
                "User",
                "Query",
            ])
            .style(Style::new().bold().dim()),
        )
        .row_highlight_style(highlight_style);

    frame.render_stateful_widget(table, area, &mut app.blocks_locks_state);
}

fn draw_detail(app: &App, frame: &mut Frame, area: Rect) {
    let [_, panel_area] =
        Layout::vertical([Constraint::Percentage(40), Constraint::Percentage(60)]).areas(area);

    let (title, query, extra_lines) = match app.blocks_active_pane {
        BlocksPane::Tree => {
            let Some(node) = app.selected_block_node() else {
                return;
            };
            let lock_info = if let (Some(m), Some(r)) = (&node.mode, &node.relation) {
                if !r.is_empty() {
                    format!("{m} on {r}")
                } else {
                    m.clone()
                }
            } else if let Some(m) = &node.mode {
                m.clone()
            } else {
                "-".to_string()
            };
            let title = format!(
                "Detail \u{2014} PID {} \u{2502} {}@{} \u{2502} {} \u{2502} {} (Enter/Esc: close)",
                node.pid,
                node.usename,
                node.application_name,
                node.state,
                format::human_duration(node.duration_secs)
            );
            let mut lines = Vec::new();
            if node.is_deadlock {
                lines.push(
                    Line::from("DEADLOCK DETECTED IN THIS CYCLE").style(Style::new().fg(Color::Red).bold()),
                );
            } else if node.is_root {
                lines.push(
                    Line::from(format!(
                        "ROOT BLOCKER \u{2014} blocking {} descendant session(s)",
                        node.num_descendants
                    ))
                    .style(Style::new().fg(Color::Red).bold()),
                );
            }
            lines.push(Line::from(vec![
                Span::styled("Awaiting lock: ", Style::new().dim()),
                Span::styled(lock_info, Style::new().cyan()),
            ]));
            (title, node.query.clone(), lines)
        }
        BlocksPane::Locks => {
            let Some(lock) = app.selected_active_lock() else {
                return;
            };
            let title = format!(
                "Detail \u{2014} Lock on PID {} \u{2502} {} \u{2502} {} (Enter/Esc: close)",
                lock.pid,
                lock.locktype,
                if lock.granted { "GRANTED" } else { "WAITING" }
            );
            let mut lines = Vec::new();
            lines.push(Line::from(vec![
                Span::styled("Mode: ", Style::new().dim()),
                Span::raw(lock.mode.clone()),
                Span::styled(" \u{2502} Target: ", Style::new().dim()),
                Span::styled(
                    if lock.relation.is_empty() {
                        "-".to_string()
                    } else {
                        lock.relation.clone()
                    },
                    Style::new().cyan(),
                ),
                Span::styled(" \u{2502} Duration: ", Style::new().dim()),
                Span::raw(format::human_duration(lock.duration_secs)),
                Span::styled(" \u{2502} User: ", Style::new().dim()),
                Span::raw(lock.usename.clone()),
            ]));
            (title, lock.query.clone(), lines)
        }
    };

    let mut lines = extra_lines;
    lines.push(Line::default());
    lines.extend(sql::highlight_lines(&query));

    let panel = Paragraph::new(lines)
        .wrap(Wrap { trim: false })
        .block(Block::bordered().title(title));
    frame.render_widget(Clear, panel_area);
    frame.render_widget(panel, panel_area);
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::Terminal;
    use ratatui::backend::TestBackend;

    #[test]
    fn blocks_lens_renders_tree_and_locks_from_mock() {
        let mut app = App::new();
        app.active_tab = crate::app::Tab::BlocksLens;

        let backend = TestBackend::new(140, 40);
        let mut terminal = Terminal::new(backend).expect("test terminal");
        terminal
            .draw(|frame| draw(&mut app, frame, frame.area()))
            .expect("draw");

        let screen: String = terminal
            .backend()
            .buffer()
            .content()
            .iter()
            .map(|cell| cell.symbol())
            .collect();

        assert!(screen.contains("Blocking Tree"));
        assert!(screen.contains("Active Locks"));
        // Mock snapshot has root blocker 4312
        assert!(screen.contains("4312"), "{screen}");
        assert!(screen.contains("ROOT BLOCKER"), "{screen}");
        assert!(screen.contains("WAIT"), "{screen}");
    }

    #[test]
    fn blocks_lens_detail_panel_renders() {
        let mut app = App::new();
        app.active_tab = crate::app::Tab::BlocksLens;
        app.detail_open = true;
        app.blocks_tree_state.select(Some(0));

        let backend = TestBackend::new(140, 40);
        let mut terminal = Terminal::new(backend).expect("test terminal");
        terminal
            .draw(|frame| draw(&mut app, frame, frame.area()))
            .expect("draw");

        let screen: String = terminal
            .backend()
            .buffer()
            .content()
            .iter()
            .map(|cell| cell.symbol())
            .collect();

        assert!(screen.contains("Detail \u{2014} PID 4312"), "{screen}");
        assert!(screen.contains("ROOT BLOCKER"), "{screen}");
    }
}
